### PR TITLE
Azure discovery plugin

### PR DESCRIPTION
**Pull request by aslanbekirov**
Please Check also https://github.com/crate/elasticsearch/pull/87

### CODE REVIEW

File: es-cloud-azure/build.gradle [hunk 1]
```diff
@@ -0,0 +1,69 @@
...
+            parentDir.mkdirs();
+        }
+        file.createNewFile();
+    }
+    ant.propertyfile(file:file.path, comment:'Plugin Descriptor') {
+        ant.entry(key:'name', value:project.archivesBaseName)
+        ant.entry(key:'description', value:project.description)
+        ant.entry(key:'classname', value:"org.elasticsearch.plugin.cloud.azure.CloudAzurePlugin")
+        ant.entry(key:'version', value:"0.1.0")
+        ant.entry(key:'elasticsearch.version', value:"2.3.4")
```
Comment (mfussenegger) [inline es-cloud-azure/build.gradle]: I think the `versions.elasticsearch` property can be used here

**Comment by mfussenegger** [comment]
Are there any distribution tests to make sure the things in this PR actually work?

**Comment by aslanbekirov** [comment]
@mfussenegger I created a jenkins job (https://jenkins.crate.io/job/azure-plugin-test/) that is creating 3 node on azure and tests azure discovery plugin. I hope this is what you meant.
In addition when I am using `versions.elasticsearch` it is compaining that "elastic search version is not specified", that is why I used `elasticsearch.version`.

**Comment by mfussenegger** [comment]
lgtm, but maybe let @seut take a look at it as well

**Comment by aslanbekirov** [comment]
thank you @mfussenegger , I will ask him, too

File: blackbox/docs/azure_discovery.txt [hunk 2]
```diff
@@ -0,0 +1,77 @@
...
+discovery in an Azure environment.
+
+Requirements to use Azure api for node discovery
+------------------------------------------------
+
+The mechanisms for discovery on Microsoft Azure are built
+in with Crate, but the required sdk libraries are not included in
+any Crate distribution and needed to be added to Crate's
+libs folder. By default this is
+``$CRATE_HOME/lib``
```
Comment (seut) [inline blackbox/docs/azure_discovery.txt]: no don't do this. we have a plugin folder for this, so the right way is to install them in `plugins/cloud-azure/libs` folder. never move plugin related libs into the global lib path, if some want to remove azure support, you must now all related libs otherwise which is kind a wrong... thats why we support loading libs from `plugins/<plugin</libs`

**Comment by chaudum** [comment]
@dobe can you take a look please

File: blackbox/docs/azure_discovery.txt [hunk 3]
```diff
@@ -0,0 +1,77 @@
+.. highlight:: yaml
+.. _azure_discovery:
+
+===============
+Azure Discovery
+===============
+
+When running Crate on Microsoft Azure you usually face
+the problem that Crate's default discovery mechanism does
+not work out of the box.
```
Comment (dobe) [inline blackbox/docs/azure_discovery.txt]: would rephrase this

File: blackbox/docs/azure_discovery.txt (hunk 2 above)
Comment (dobe) [inline blackbox/docs/azure_discovery.txt]: update docs

File: es/upstream [hunk 4]
```diff
@@ -1 +1 @@
-Subproject commit 4e547cb7fa176eeb9710d021fbbc13321f8d0b72
+Subproject commit ead2009993cc23312b16321265d0a8fe9c25cfff
```
Comment (dobe) [inline es/upstream]: do we still have changes here which are required?

File: app/build.gradle [hunk 5]
```diff
@@ -331,15 +330,27 @@ jar {
...
+    // copy jar
+    copy {
+        from(project(':azure-discovery').tasks.jar.archivePath)
+        into 'plugins/crate-azure-discovery'
+    }
+
+    // copy generated plugin-descriptor.properties
+    copy {
+        from(project(':azure-discovery').buildDir.path + '/tmp')
+        into 'plugins/crate-azure-discovery'
```
Comment (dobe) [inline app/build.gradle]: should work with a single copy

File: azure-discovery/src/main/java/io/crate/azure/discovery/AzureUnicastHostsProvider.java [hunk 6]
```diff
@@ -0,0 +1,282 @@
...
+        @Override
+        protected List<DiscoveryNode> refresh() {
+            ArrayList<DiscoveryNode> nodes = new ArrayList<>();
+            InetAddress ipAddress = null;
+            try {
+                ipAddress = networkService.resolvePublishHostAddresses(null);
+                logger.trace("ip of current node: [{}]", ipAddress);
+            } catch (IOException e) {
+                // We can't find the publish host address... Hmmm. Too bad :-(
+                logger.trace("exception while finding ip", e);
```
Comment (dobe) [inline azure-discovery/src/main/java/io/crate/azure/discovery/AzureUnicastHostsProvider.java]: in this case there will be probably an NPE below. should probably return empty here?
also i think this shuld be at least WARN level

File: blackbox/docs/azure_discovery.txt [hunk 7]
```diff
@@ -53,25 +51,44 @@ Please add the following libraries into the libs folder::
     - lang-tag-1.4.jar
     - nimbus-jose-jwt-3.1.2.jar
     - oauth2-oidc-sdk-4.5.jar
-    - stax-api-1.0-2.jar
```
Comment (chaudum) [inline blackbox/docs/azure_discovery.txt]: this is already included in crate

**Comment by dobe** [comment]
changelog

File: azure-discovery/src/main/java/io/crate/azure/AzureModule.java [hunk 8]
```diff
@@ -0,0 +1,133 @@
...
+    public static void registerServices(Registry registry) {
+        // taken from https://github.com/Appdynamics/azure-connector-extension/blob/master/src/main/java/com/appdynamics/connectors/azure/ConnectorLocator.java
+        new com.microsoft.windowsazure.core.pipeline.apache.Exports().register(registry);
+        new com.microsoft.windowsazure.core.pipeline.jersey.Exports().register(registry);
+        new com.microsoft.windowsazure.core.utils.Exports().register(registry);
+        new com.microsoft.windowsazure.credentials.Exports().register(registry);
+        new com.microsoft.windowsazure.management.configuration.Exports().register(registry);
+        new com.microsoft.azure.management.compute.Exports().register(registry);
+        new com.microsoft.azure.management.storage.Exports().register(registry);
+        new com.microsoft.azure.management.network.Exports().register(registry);
```
Comment (chaudum) [inline azure-discovery/src/main/java/io/crate/azure/AzureModule.java]: @dobe  registering the azure services manually did the job ...

**Comment by chaudum** [comment]
since @dobe is on holidays, @seut you will need to finalise the review of this PR

File: azure-discovery/src/main/java/io/crate/azure/management/AzureComputeServiceImpl.java [hunk 9]
```diff
@@ -111,14 +113,13 @@ private Configuration createConfiguration() {
...
-                authRes.getAccessToken()
-            );
+
+            DefaultBuilder registry = DefaultBuilder.create();
+            AzureModule.registerServices(registry);
+            conf = ManagementConfiguration.configure(null, new Configuration(registry),
+                URI.create(Azure.ENDPOINT), subscriptionId, authRes.getAccessToken());
         } catch (Exception e) {
-            logger.error("Could not create configuration for Azure clients: {}", e.getMessage());
+            logger.error("Could not create configuration for Azure clients: {}", e);
```
Comment (seut) [inline azure-discovery/src/main/java/io/crate/azure/management/AzureComputeServiceImpl.java]: remove place holder as it is not used (exception is passed in here for complete stacktrace).

**Comment by seut** [comment]
lgtm
